Fontawesome support

jeabakker opened the pull request:
wip for #7239 

I replaced the Elgg sprites with FontAwesome icons, please give feedback
- [x] update docs
- [x] check hard coded sprite refs
- [x] serve font-awesome.css via simplecache

mrclay commented on engine/lib/output.php:
```diff
@@ -468,6 +468,67 @@ function elgg_strip_tags($string, $allowable_tags = null) {
+ *
+ * @return string
+ */
+function elgg_convert_elgg_icon_to_fontawesome_icon($elgg_icon_class) {
```
We don't want this as public API right?

jeabakker replied:
probably not but i don't know the policy for calling private functions in (public) views

hypeJunction commented:
I am not a big fan of this approach. It introduces a new vendor prefix that we do not need. It should be easy to switch to any other font

mrclay commented:
> It should be easy to switch to any other font

@hypeJunction To do this I guess we'd need to allow devs to add their own mapping like `elgg_convert_elgg_icon_to_fontawesome_icon` does here?

hypeJunction commented:
@mrclay FontAwesome naming conventions are pretty awesome. One way would be use less/sass to compile an FA stylesheet with a custom prefix (.elgg-icon-). Then themes can remap to their own glyphs.

jeabakker commented:
> I am not a big fan of this approach. It introduces a new vendor prefix that we do not need. It should be easy to switch to any other font

it is, just take over 1 view and use your own icon library


> One way would be use less/sass to compile an FA stylesheet with a custom prefix (.elgg-icon-). Then themes can remap to their own glyphs.

But then we (Elgg) have to maintain that stylesheet, now we can just use composer to keep the icons up-to-date

hypeJunction commented:
> it is, just take over 1 view and use your own icon library

Then don't add .fa styles to the stylesheet?

hypeJunction commented:
And you are adding new API. Perhaps using a array map in the view would suffice?

hypeJunction commented on mod/aalborg_theme/views/default/aalborg_theme/css.php:
```diff
@@ -67,6 +67,22 @@
+.elgg-icon-hover-menu,
+.elgg-icon-hover-menu:hover,
+:focus > .elgg-icon-hover-menu {
+	background: transparent url(<?= elgg_get_simplecache_url('aalborg_theme/avatar_menu_arrows.png') ?>) no-repeat;
```
We've dropped support for older browsers. Can't we just use a CSS triangle?

jeabakker replied:
i think this was added so mobile users could click on the user icon, maybe @PerJensen can tell us if this is still needed?

hypeJunction commented on views/default/css/admin.php:
```diff
@@ -1300,42 +1300,43 @@
 /* ***************************************
 	ICONS
 *************************************** */
+.fa {
```
IMHO, can use .elgg-icon and drop the font-size (it should inherit the font-size of the text)

jeabakker replied:
i'll look in to removing all .fa css from the core files and replace it with .elgg-icon

hypeJunction commented on views/default/css/admin.php:
```diff
@@ -1300,42 +1300,43 @@
+h4 > .fa,
+h5 > .fa,
+h6 > .fa {
+	font-size: 1em;
```
1em is 16px in a default setup. I would say let's use:

```css
* > .elgg-icon {
  color: #aaaaaa;
  font-size: 1em;
  line-height: 1em;
  vertical-align: middle;
}
```

jeabakker replied:
in Elgg 1em = 14.4px because of https://github.com/Elgg/Elgg/blob/master/views/default/elements/typography.css.php#L15

hypeJunction commented:
Also, let's not duplicate the icons in admin CSS and just load them by default.
We can also map the vendor css to a view, and just extend ```css/elements/icons```

jeabakker committed:
feature(icons): use FontAwesome icons

BREAKING CHANGE:
If you use a reference to the physical Elgg sprites please update your
plugin to use the FontAwesome icons.

Replace the Elgg sprites icons by FontAwesome icons, this allowes for
more icons and easier styling of the icons

fixes #7239

jeabakker committed:
chore(icons): update icons css to support FontAwesome

jeabakker commented:
rebased, added a section in docs and applied some of the suggestions

hypeJunction commented:
Thanks. LGTM

jdalsem commented:
Ready to merge now @jeabakker?

ewinslow commented on engine/lib/views.php:
```diff
@@ -1519,6 +1519,9 @@ function elgg_views_boot() {
 	elgg_register_js('elgg', elgg_get_simplecache_url('elgg.js'), 'head');
 	elgg_load_js('elgg');
+	
+	elgg_register_css('fontawesome', 'vendor/fortawesome/font-awesome/css/font-awesome.min.css');
```
Don't point to the vendor file directly. Instead, register it as a view with `engine/views.php` and then point to the simplecache asset. View should be "font-awesome.css"

ewinslow commented on mod/developers/views/default/theme_sandbox/icons.php:
```diff
@@ -3,7 +3,7 @@
  */
 
-echo elgg_view_module('theme-sandbox-demo', 'Icon Sprites (.elgg-icon)', elgg_view('theme_sandbox/icons/sprites'));
+echo elgg_view_module('theme-sandbox-demo', 'Icons (.elgg-icon)', elgg_view('theme_sandbox/icons/sprites'));
```
Thank you for removing any trace of this implementation detail...

ewinslow commented on views/default/admin.css.php:
```diff
@@ -1381,47 +1379,6 @@
 	width: 200px;
 	height: 200px;
 }
-.elgg-avatar > .elgg-icon-hover-menu {
```
Why was the elgg-avatar stuff deleted?

jeabakker replied:
duplicate css ;)

ewinslow commented on views/default/output/icon.php:
```diff
@@ -1,18 +1,84 @@
+* @uses $vars['convert'] Convert an elgg-icon class to a FontAwesome class (default: true)
+*/
+
+// these 'old' Elgg 1.x sprite icons will be converted to the FontAwesome version
```
Wow, nice work with BC!

jeabakker commented:
registered the css as a view in `engine/views.php` as suggested by @ewinslow

mrclay commented:
See #8625
